pkg/operator/controllers/machineset/machineset_controller.go

```diff
@@ -0,0 +1,84 @@
+package machineset
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+)
+
+type Reconciler struct {
+	log *logrus.Entry
+
+	arocli aroclient.Interface
+	maocli maoclient.Interface
+}
+
+// MachineSet reconciler watches MachineSet objects for changes, evaluates total worker replica count, and reverts changes if needed.
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, maocli maoclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:    log,
+		arocli: arocli,
+		maocli: maocli,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.Features.ReconcileMachineSet {
+		return reconcile.Result{}, nil
+	}
+
+	machinesets, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	modifiedMachineset, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Count amount of total current worker replicas (including custom MachineSets, if present)
+	replicaCount := 0
+	for _, machineset := range machinesets.Items {
+		if machineset.Spec.Replicas != nil {
+			replicaCount += int(*machineset.Spec.Replicas)
+		}
+	}
+
+	if replicaCount < minSupportedReplicas && strings.Contains(modifiedMachineset.Name, instance.Spec.InfraID) {
```

Comment:
The controller won't intervene in this case as long as total worker replica count is 3, because the List operation here counts all worker replicas towards replica count, not just ones matching our infra ID: https://github.com/cadenmarchese/ARO-RP/blob/machineset-controller/pkg/operator/controllers/machineset/machineset_controller.go#L49

If the user were to scale down our replicas before making their own, then the controller would intervene.

Reply:
@cadenmarchese yes, I think the controlelr won't entervene in the above case, but the customer ends up with 3 masters and 0 workers after the step 4. No?

Reply:
You're right, the controller wouldn't scale ours back up, because they aren't the objects being modified at that time. I am struggling to think of how to account for this use case without changing other assumptions (for example, that custom machinesets should count, and that we aren't modifying custom objects). I'm happy for suggestions.

Reply:
@m1kola if we go with option 2 and introduce webhooks to block machineset deletion later (as has been discussed) the customer would be stuck with our machinesets. For this reason I would probably recommend the first option + some added documentation for customers regarding the feature.

I actually don't think option 2 would be too hard, we could just filter the machineset List by our InfraID (@troy0820 suggested this). For option 1 we could just remove all the `strings.Contains` comparisons and just evaluate/edit all objects.

Reply:
> if we go with option 2 and introduce webhooks to block machineset deletion later

I would prefer to avoid using webhooks, if possible. At least for now. If I understand correctly description of work item 8749435, it suggests that we do the second option and if our default machine sets no longer exist at the moment of reconciliation - we just give up. Cluster is unsupported. I would start with this and see if it improves the situation. If not - we can reconsider webhooks.

> For option 1 we could just remove all the strings.Contains comparisons and just evaluate/edit all objects.

Sounds correct

> I actually don't think option 2 would be too hard, we could just filter the machineset List by our InfraID

It is a bit more complicated than this. In multi-az setup we have 3 machine sets (1 for each region). Imagine the scenario when:

1. Customer creates a new cluster with 3 masters and 3 worker nodes in multi az region (meains 3 machine sets each having 1 replica).
2. Customer creates a new custom machine set with 2 replicas
3. Customer scales default machine sets to 0. At this point we are still good - we still have 2 workers
4. Customer deletes custom machine set.  At this point we have 3 masters and 0 workers, but 3 default machine sets. 

What do we do now? Listing all machine sets with infra id and scaling one of them to 2 replicas? I don't think it is right approach because the cluster ends up with worker nodes in single availability zone while region has a support of multiple azs.

IMO we should scale 2 machine sets to have 1 replica each in this situation.

We also need to keep in mind:

* Single AZ regions (one default machine set with 3 replicas)
* Case when a customer has a cluster in multi az region, but deleted some one or two of the default machine sets
* As mentioned before, when customer deletes all default machine sets - we give up.

What do you think?

Reply:
It sounds like the path forward is:

- Continue to watch all MachineSets for changes, and continue to evaluate all MachineSets towards replicaCount
- Determine if AZ or not (3 default MachineSets versus 1 default MachineSet)
   - If it is AZ, and reconcile finds less than 3 replicas, add 1 to each default MachineSet (unless already 1)
   - If it isn't AZ, and reconcile finds less than 3, add minSupportedReplicas - replicaCount back to the single default MachineSet
   - If `IsNotFound`, bail out.

The assumption I'm making above is that we want 3 replicas (the user story said 2 initially, but MSFT support policy says 3).

Reply:
@cadenmarchese right, we need 3, not 2. It used to be 2, but we increased the requirement. I got confused by looking at the work item and my previous comment assumes that we need 2. Could you please update the work item to reflect the new requirements?

> Determine if AZ or not (3 default MachineSets versus 1 default MachineSet)

This not always going to be the case: customer can remove one of three machine sets with infra id. I think the logic goes down to this:

* Continue to watch all MachineSets for changes, and continue to evaluate all MachineSets towards replicaCount
* If `IsNotFound` while getting machine set which is being reconciled (`request.Name`), bail out.
* If `replicaCount < 3`, then:
    * Find all remaining machine sets with infra id
    * Sort them by number of replicas: from lower to higher
    * Increase each each until we hit `replicaCount == minSupportedReplicas`.

I'm thinking about somthing like this (but with actual machine set objects):

```go
mSetsWithInfraId := []int{0, 0, 0}
replicaCount := 0
const wantReplicas = 3

sort.Sort(sort.IntSlice(mSetsWithInfraId))

for i := 0; replicaCount < wantReplicas && len(mSetsWithInfraId) > 0; {
	mSetsWithInfraId[i]++
	replicaCount++

	i++
	if i >= len(mSetsWithInfraId) {
		i = 0
	}
}

fmt.Printf("%#v\n", mSetsWithInfraId)
```

So results will like this:


| initial `mSetsWithInfraId`  | resulting `mSetsWithInfraId` | Comment |
| ------------------------------ | ---------------------------------- | --------- |
| []int{0}  | []int{3}  | Most common case without AZ support  |
| []int{0, 0}  | []int{2, 1}  | Customer deleted one of the machin sets with infra id |
| []int{0, 0, 0} | []int{1, 1, 1} | most common case with AZ support |
| []int{1, 0, 0} | []int{1, 1, 1} | initially `replicaCount == 1` |
| []int{2, 0, 0} | []int{1, 0, 2} | initially `replicaCount == 2` |
| []int{3, 0, 0} | []int{1, 0, 2} | initially `replicaCount == 3`, no action from the controller |
| []int{1, 1, 0} | []int{1, 1, 1} | initially `replicaCount == 2` |

Reply:
But I want to take a step back and look at the first option:

> Scale customer created machine sets back to meet the requiremenrt (technically easisest solution, but might be more confusing to the customer)

Wondering what are the other implications of this apart from being confusing. Maybe we should just do this?

@mjudeikis what is your take on this topic?

Reply:
I think our goal was to keep it simple, and bailout in any way we deviate from default configuration.
1. If customer deleted original workerSets - bailout
2. If customer scaled down original machineSet AND where is no other machinesets available - scale up.
3. If customer scaled down original machineSet AND where is other machienset (dont care Az or available replicas) - bail out as we can't cover ALL customer ideas and thinking behind.

So in theory we covering only basic scenario where customer didnt do any magic in MachineSet just got idea to scale down existing one.

Reply:
@mjudeikis so what you are saying is: only watch for original machine set and try to scale up if needed and ignore everything else, right?

Reply:
> So in theory we covering only basic scenario where customer didnt do any magic in MachineSet just got idea to scale down existing one.

From comments above, I am understanding:

- We bail out if any custom machinesets are found in the list of machinesets (we can't cover all use cases here, so assume positive intent and leave it alone)
- We watch for changes to machineset objects and revert them if needed (no need to do AZ sorting, because we are just watching for scale down)
- We are moving to 2 minSupportedReplicas instead of 3 (https://github.com/Azure/ARO-RP/pull/1532#discussion_r678178634), which is fine in the context of our upgrades and the user story

Does this sound accurate @m1kola @mjudeikis?

Reply:
> We bail out if any custom machinesets are found in the list of machinesets (we can't cover all use cases here, so assume positive intent and leave it alone)

No. My undersntanding of MJ's idea is that we should do this:

* If reconciliation request is for default worker node (with infra Id) we proceed with reconciliation and check total number of worker nodes
* If the request is for something else - we bail out

Everything else sounds good to me, but I would like MJ to confirm that we understood his idea correctly.

Reply:
These two parts:

> bailout in any way we deviate from default configuration

> If customer scaled down original machineSet AND where is other machienset (dont care Az or available replicas) - bail out

To me, this sounds like we are bailing out in any case where there are custom machinesets (not just when a custom machineset triggers reconcile). Something like this:

```
for _, machineset := range machinesets.Items {
	if !strings.Contains(machineset.Name, instance.Spec.InfraID)) {
		return reconcile.Result{}, nil // don't do anything, and don't requeue
	}
	if machineset.Spec.Replicas != nil {
		replicaCount += int(*machineset.Spec.Replicas)
	}
}
```
So we would still reconcile each time an object is changed, but this way we're covered if the second point above is true. Let me know if I am not understanding correctly.

Reply:
I caught up with MJ about this, and we want to bail out in either case. @m1kola do you mind reviewing the controller code changes? From there we can wrap up tests.